### PR TITLE
Fix panic on network timeout during push

### PR DESCRIPTION
**Pull request by aaronlehmann**
`Upload` already closes the reader returned by `compress` and the progressreader passed into it, before returning. But even so, the io.Copy inside compress' goroutine needs to attempt a read from the progressreader to notice that it's closed, and this read has a side effect of outputting a progress message. If this happens after `Upload` returns, it can result in a write to a closed channel. Change `compress` to return a channel that allows the caller to wait for its goroutine to finish before freeing any resources connected to the reader that was passed to it.

The underlying bug is not new in 1.10, but switching to a channel for progress output exposes this as a panic instead of having a write fail silently after the HTTP writer is closed. Suggest 1.10 milestone.

cc @jfrazelle

### CODE REVIEW

**Comment by icecrime** [comment]
Thanks @aaronlehmann! Adding to the milestone indeed (cc @tiborvass @thaJeztah)

**Comment by tiborvass** [comment]
LGTM

**Comment by thaJeztah** [comment]
> Adding to the milestone indeed

SGTM, thanks!

File: distribution/push_v2.go [hunk 1]
```diff
@@ -345,8 +345,11 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
 	size, _ := pd.layer.DiffSize()
 
 	reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, arch), progressOutput, size, pd.ID(), "Pushing")
-	defer reader.Close()
-	compressedReader := compress(reader)
+	compressedReader, compressionDone := compress(reader)
+	defer func() {
+		reader.Close()
+		<-compressionDone
```
Comment (MHBauer) [inline distribution/push_v2.go]: If we're waiting for the compressionDone channel to close, doesn't it need to happen before something else? I don't see how this delays anything at the end of a deferred function. I'm probably misinterpreting this.

Comment (aaronlehmann) [inline distribution/push_v2.go]: The channel close happens after this function returns, here: https://github.com/docker/docker/blob/master/distribution/xfer/download.go#L199

Comment (MHBauer) [inline distribution/push_v2.go]: I am out of my league.

**Comment by cpuguy83** [comment]
LGTM pending janky... last run had different failures on each one...

**Comment by thaJeztah** [comment]
added cherry-pick label

**Pull request merged**